set-java-home.bash

```diff
@@ -1,19 +1,20 @@
 asdf_update_java_home() {
   local java_path
-  java_path="$(asdf which java)"
+  java_path="$(asdf where java)"
```

Comment:
This might be ok - the reason we switched to `where` is because it behaves differently when the jdk is set to `system`.

```bash
  % asdf where java
System version is selected
  % asdf which java                                                                                                                                     
usr/bin/java
```

Reply:
It seems like the call to `asdf` itself is what's slowing things down, so using the `where` command suffers the same lag.  I wonder if there is some sort of callback that we can subscribe to when `asdf` identifies that it needs to use a new `.tools-version`.

Reply:
Just to follow up on my last train of thought.  The callback option may not actually work because the tool versions are "not known" until the shims are actually executed.  So when calling something like `java -version`, `asdf` doesn't know which installed Java to use until at that very moment.  If we were saving the last known Java version somewhere, we could potentially be out of sync because we may have traversed to a different folder already.  A callback would not be able to keep the JAVA_HOME up-to-date in the same manner. 

So to summarize, this is a pretty tough nut to crack.  @smorimoto reviewed my PRs in the past; maybe they could weigh in?

Reply:
FYI if you set `asdf global java system` then $JAVA_HOME gets incorrectly set. I think it should skip setting `JAVA_HOME` or something if system is provided.

Reply:
I had to update this to work when no java is available otherwise `JAVA_HOME` would get set to whatever `asdf where java` was outputting.

```bash
asdf_update_java_home() {
	local java_path
	if (asdf where java >/dev/null 2>&1); then
		java_path="$(asdf where java)"

		if [[ -n "${java_path}" ]]; then
			export JAVA_HOME
			JAVA_HOME="${java_path}"
		fi
	fi
}
```

Reply:
@jsmestad / @halcyon :
This may be a stupid question and I apologise if it is why are you exporting JAVA_HOME _before_ setting it rather than after?